### ISSUE TITLE
Meteor 1.7.1-rc.11 slower than rc.10 due to Node 8.12.0 update.

### ISSUE THREAD

**Issue by p3pp8** [meteor/meteor#10216]
I've noticed a huge slowdown of my app after upgrading to RC.11 release from RC.10.

**Comment by abernix** [comment]
Ref: https://github.com/meteor/meteor/pull/9942#issuecomment-420669483

**Comment by benjamn** [comment]
cc @nick-gudumac @veered

**Comment by veered** [comment]
A few questions:
(a) Does your app now have high CPU usage?
(b) Did your app (before the upgrade) already have high CPU usage? 
(c) How are you measuring the performance issues?

It would be extremely helpful if you take a CPU profile before & after the Node upgrade using https://github.com/qualialabs/profile

FWIW, Qualia has been running a Node 8.12 RC in production for a few weeks and we haven't noticed any performance penalty. However, we aren't using Meteor 1.7.1, we are still using 1.6.0.1 until 1.7.1 is officially released.

**Comment by macrozone** [comment]
Maybe a stupid question, but do you really run on node 8.12.0 on production? On galaxy, this is surely the case, but if you - like me - deploy on a custom environment, this is not guaranteed as node is not part of meteor's bundle.

Do the slowdowns also occur on develop?

**Comment by KoenLav** [comment]
@macrozone very good point!

**Comment by macrozone** [comment]
@jimrandomh as you opened the issue https://github.com/meteor/meteor/issues/10117  it would also be interesting to hear about your observation with node 8.12.

could https://github.com/meteor/meteor/issues/9796 be resolved with 8.12. or was the fiber spike issue not related to that issue at all?

**Comment by nick-gudumac** [comment]
In development and local build I haven't seen anything unusual, just when deployed to production, the CPU spiked like crazy. I suspect it being related to Fibers / Oplog.

@benjamn I followed your instructions and did a CPU profiling on local build. I haven't seen anything outstanding. Check them out below. Let me know if you need more information.

1.7.1 RC.11 build (Node v8.12.0)

[CPU-20180913T165128-RC11-proper.cpuprofile.zip](https://github.com/meteor/meteor/files/2379647/CPU-20180913T165128-RC11-proper.cpuprofile.zip)

1.7.1 RC.12 build (Node v8.11.4)
[CPU-20180913T170450-RC12-proper.cpuprofile.zip](https://github.com/meteor/meteor/files/2379659/CPU-20180913T170450-RC12-proper.cpuprofile.zip)

**Comment by benjamn** [comment]
@nick-gudumac Ah yes, those profiles don't seem to be showing anything suspicious.

Maybe you could try using `qualia:profile` from `meteor shell` in production? https://github.com/qualialabs/profile#runtime-profiling

**Comment by nick-gudumac** [comment]
@benjamn unfortunately I can't ssh into Galaxy, is there any other way to test this?

**Comment by sebakerckhof** [comment]
@nick-gudumac , if you can reproduce it when spawning a second project with the same code base, you could try: https://github.com/qualialabs/web-shell by cloning it manually and removing this line: https://github.com/qualialabs/web-shell/blob/master/package.js#L7 . But you obviously don't want to push this to your production environment.

**Comment by veered** [comment]
To get a profile in production, you could also create a route (with a random, secure name) that triggers a profile and then returns the file so you can download it. Maybe I can try to whip something like that up this weekend, but you could also easily do it on your own.

Pretty rough that people using Galaxy can't use prod-shell... it's not like allowing SSH into Galaxy containers would change the security model much. The Meteor server running in the container can already run arbitrary shell commands.

**Comment by MylesBorins** [comment]
Please let me know if we caused major regressions  in 8.12.0. I can't think of anything that stands out that could have caused it but we can easily revrt something and cut an 8.12.1

Please lmk if we need to bisect and I can help

**Comment by benjamn** [comment]
@nick-gudumac Any luck with @veered's suggestion? You could also use a `Meteor.call` method to start the profile, and restrict it to a specific user (i.e. you, or any admin).

**Comment by nick-gudumac** [comment]
@benjamn I've managed to take a CPU profile in a production environment under load, below are the results. Can you take a look?

[cpu-profile-meteor-1.7.1-RC-11.cpuprofile.zip](https://github.com/meteor/meteor/files/2394043/cpu-profile-meteor-1.7.1-RC-11.cpuprofile.zip)

Let me know if you need more information, or if I need to take another CPU profile.

**Comment by benjamn** [comment]
Thanks @nick-gudumac!

Wow, a ton of time is being spent in the garbage collector! 87% of the profile, or 9.6 out of 11 seconds. Nowhere close to 11 seconds of work is happening here, but everything in the chart seems to bottom out in long GC pauses. There's a reasonable diversity of callers, too, so I can't blame this on (say) Kadira, though it does show up in this profile.

**Comment by veered** [comment]
@nick-gudumac It's worth checking that using the profiler didn't increase the memory usage on the server (aka the mem usage was roughly the same both before and after running the profiler). 

This is only relevant if the profile you uploaded wasn't the first one you triggered since the server booted up. I've seen the profiler increase mem usage a couple times. If your server was already close to running out of memory, then it's possible running the profiler put it over the edge.

If the mem or CPU usage didn't increase after running the profiler or if your uploaded profile was the first profile you recorded, than the profiler isn't the issue.

**Comment by nick-gudumac** [comment]
@benjamn Yeah, I've seen the long gc pauses, any idea why this might happen? How can we trace this more, perhaps try different configurations for the garbage collector?

@veered the above profiling was not taken at server startup but initiated via method call. There was almost no memory usage change during the profiling or else, and the profiler did not cause any CPU spikes alone. Seem like the spikes are created by subscriptions and method calls (even a few calls), and the app becomes unresponsive after that for a while.

**Comment by csillag** [comment]
Since this issue seems to be blocking a critical path leading to a major performance improvement in our use case, may I ask who has the ball on this one? Thank you.

**Comment by benjamn** [comment]
@csillag Is your performance improvement related to Node 8.12.0, or just general Meteor 1.8 improvements (or both)?

**Comment by csillag** [comment]
We are waiting for Node 8.12.0, for the Fiber fix.

**Comment by MylesBorins** [comment]
@benjamn have you figured out if the GC stuff is related to a change in 8.12.0? Here are the changes that landed on V8 in that version

```
:...skipping...
* 5294919d05 - deps: V8: cherry-pick 9040405 from upstream (5 weeks ago)<Junliang Yan>
* ae63db8624 - deps: backport 804a693 from upstream V8 (6 weeks ago)<Matheus Marchini>
* d9ab189f55 - deps: cherry-pick b767cde1e7 from upstream V8 (6 weeks ago)<Ben Noordhuis>
* 31883368c7 - deps: cherry-pick 0c35b72 from upstream V8 (6 weeks ago)<Gus Caplan>
* ffb72f810e - deps: cherry-pick 09b53ee from upstream V8 (6 weeks ago)<Anna Henningsen>
* 8e0f28b8f0 - deps: V8: backport 49712d8a from upstream (6 weeks ago)<Ali Ijaz Sheikh>
* efe28b8581 - deps: V8: fix bug in InternalPerformPromiseThen (6 weeks ago)<Ali Ijaz Sheikh>
* 9aeffab452 - deps: V8: cherry-pick 8361fa58 from upstream (6 weeks ago)<Ali Ijaz Sheikh>
* f987a512d4 - deps: V8: backport b49206d from upstream (6 weeks ago)<Ali Ijaz Sheikh>
* e6cd7e57b3 - deps: V8: cherry-pick 5ebd6fcd from upstream (6 weeks ago)<Ali Ijaz Sheikh>
* d868eb784c - deps: V8: cherry-pick 502c6ae6 from upstream (6 weeks ago)<Ali Ijaz Sheikh>
* 656ceea393 - deps: cherry-pick dbfe4a49d8 from upstream V8 (6 weeks ago)<Jan Krems>
```

**Comment by benjamn** [comment]
@csillag What if we put out Meteor 1.8 with Node 8.11.4 (same as currently), and then immediately release a meteor 1.8.1 with Node 8.12.0, without "recomending" that release until we can get to the bottom of these problems, so that your team could update?

**Comment by csillag** [comment]
I'm not particularly interested in what tag we put on the release.
I would also be perfectly fine with an 1.7.2-beta.0 release, or anything you can think of.
Just include the Fiber fix, please. :)

**Comment by stelund** [comment]
Seems like @p3pp8 just open this item and everyone blamed the node upgrade. I see no confirmation that he even updated to 8.12.0 for his test. After finding about the Fiber issue today which looks suspicious to what I saw in our production environment I will try to upgrade node to 8.12.0. I can't see any major bugs reported to v8.x tag for nodejs since the 16 days old release of 8.12.0.

**Comment by veered** [comment]
@stelund That's a good point, we might be jumping to conclusions. I've noticed large performance improvements across the board from 8.12.0

**Comment by nick-gudumac** [comment]
@veered have you tried node 8.12.0 with latest Meteor 1.8 RC? I'm just trying to understand if the issue is due to some packages / configurations or some regressions in node. I got excited by this update, expecting major performance improvements (and finally a fix for occasional CPU spikes), but it doesn't seem to work out.

**Comment by veered** [comment]
@nick-gudumac I've been using Node 8.12.0 w/ Meteor 1.6.0.1. I suppose that's some evidence that it is related to Meteor 1.8 rather than Node 8.12.0. 

Also, I'm a bit confused by all this, because the version of Node which runs in production has nothing to do with the version of Meteor. In production the version of Node will be whatever is installed on the server. Meteor only controls the version of Node running in development mode.

Are people with performance issues in production sure that they are actually even using Node 8.12.0?

**Comment by benjamn** [comment]
@veered Galaxy takes the Node version from the Meteor release, so anyone using Galaxy (which includes @nick-gudumac, I believe) is definitely using Node 8.12.0 in production if they deploy an app that's pinned to the Meteor release candidate that included Node 8.12.0 (namely, 1.7.1-rc.11). I don't see any reason to be skeptical of whether Node 8.12.0 was used, though I supposed it's worth double-checking that the Meteor version was indeed 1.7.1-rc.11.

**Comment by veered** [comment]
Oh good point, I was only thinking about the self-hosted case

Has anyone tried running Meteor 1.8 without Node 8.12.0?

**Comment by benjamn** [comment]
Anyone running the latest RCs on Galaxy will be using Meteor 1.8 with Node 8.11.4.

**Comment by nanu-c** [comment]
I try to use node-canvas to merge images and I stumble upon this error
`undefined symbol: _ZN2v87Isolate19CheckMemoryPressureEv`
which is a bug in the Node 8.11.4. and should be resolved in the newest version.
https://github.com/Automattic/node-canvas/issues/1252

**Comment by danielstaleiny** [comment]
@nanu-c  same issue with node-canvas. I am using it in aws lambda so update to 10/11/12 is not possible. and forcing different version of the node is also not very practical. I think it is regression.

**Comment by benjamn** [comment]
For those who have been wanting to run Node 8.12.0 in production on Galaxy, you can now update your apps to Meteor 1.8.1-beta.0 by running the following command:
```bash
meteor update --release 1.8.1-beta.0
```
If you deploy to Galaxy with this version of Meteor, please keep an eye on your CPU usage, and report your findings (good or bad) here.

**Comment by csillag** [comment]
We have been eagerly waiting for this, although we don't use Galaxy (company policy reasons). Unfortunately, we can't prove any meaningful feedback as of now, because we have hit another (independent) performance problem. (This is a live system, and data patterns are also changing.) I will report back when we have resolved that conundrum, and we are in a better position to take some measurements.

**Comment by elie222** [comment]
Just to add to this conversation from a different angle, we were experiencing memory leaks on our 1.6 app. Likely to do with our Apollo resolvers, but we didn't figure it out yet. Upon upgrading to Meteor 1.7 these memory issues seem to have stopped and memory stays around 500mb for each instance. It does hit 800mb or so occasionally, but then drops down after a while. Before the upgrade I was seeing instances randomly jump from 500mb ram usage to 2gb ram usage in a matter of hours. 

Didn't quite understand how this happened, but if the GC is doing more work now, could be that's why our memory is better controlled now.

**Comment by oohaysmlm** [comment]
Possibly posted this in the wrong place earlier, but I think we had a similar issue which I was describing over here: https://github.com/meteor/meteor/issues/10032

**Comment by KoenLav** [comment]
@erikolofsson

https://pages.mtu.edu/~shene/NSF-3/e-Book/SEMA/basics.html

It seems that garbage collection is triggered, awaiting access to a resource, which somehow causes garbage collection to run again, awaiting access to that same resource (which it is apparently never granted access to). 

@MylesBorrins I am not sure if my interpretation above is correct, would you care to shed some more light?

@benjamn am I correct in assuming this will probably have something to do with the use of Fibers?

If so then maybe @laverdet can help pinpointing the problem?

**Comment by erikolofsson** [comment]
I downgraded to node 8.11.4 and the problem persists, now with a slightly different call stack:

```

**Comment by KoenLav** [comment]
Judging from the report above it is now more likely that the error originates in Meteor or one of its dependencies, I guess.

**Comment by jimrandomh** [comment]
That sounds like memory corruption, which narrows the scope for where the bug could be to things that involve native code.

**Comment by AndrewMorsillo** [comment]
I'm also seeing problems on meteor 1.8 and node 8.12.0. If too many users connect at once my instance goes into some sort of garbage collection loop that it never recovers from. Ram usage climbs up to max_old_space_size, CPU stays pegged at 100% and 80%+ of time is spent in the GC. This only happens if lots of users connect at once however. The same amount of users connecting over a period of time will work just fine and the server will hum along without problem.

Here is a CPU profile and some screenshots to illustrate. I scaled my app down from 6 pods to 1 to trigger this problem. As soon as the users from the other pods connect to the single remaining one then it goes haywire. If I scale back up then things calm down again.

<img width="1682" alt="screen shot 2018-11-21 at 2 28 11 pm" src="https://user-images.githubusercontent.com/858965/48866100-a1817900-ed9f-11e8-8ee6-f9b2d21f0e7f.png">
<img width="1484" alt="screen shot 2018-11-21 at 2 27 39 pm" src="https://user-images.githubusercontent.com/858965/48866101-a1817900-ed9f-11e8-9715-9fac9bd5d622.png">

[2018-11-21T19_27_28.302Z.cpuprofile.zip](https://github.com/meteor/meteor/files/2605635/2018-11-21T19_27_28.302Z.cpuprofile.zip)

**Comment by laverdet** [comment]
@AndrewMorsillo do you have the ability to build fibers from source on your system? If so could you please try to rebuild from source while on nodejs v8.12.0 and see if that changes anything? You can do this by doing `cd node_modules/fibers; node build -f`.

**Comment by AndrewMorsillo** [comment]
@laverdet I can't reproduce this issue in development. I only see it in production. I suppose I could try to add a rebuild of fibers to my Dockerfile. What change would you expect to see by rebuilding fibers? Any specific issue you think that would solve that you can point me to?

**Comment by KoenLav** [comment]
@AndrewMorsillo thanks for digging into this!

What are your findings so far?

**Comment by AndrewMorsillo** [comment]
@KoenLav I seem to have achieved some stability by removing all "reactive publishing" from my app. I previously had issues with the peerlibrary:reactive-publish package so I removed that and replaced it with a custom solution using this.added and observeChanges. After a lot of fighting with attempting to take heap dumps which always crashed node I decided to modify my app to remove any of this reactive publishing. After doing so my app is much more stable. I'm running with about 50 users per instance with each instance hovering around 2gb of memory and 30% cpu. 2gb and 30% still seems a little high so I may do more digging but it also may be good enough for our purposes. 

I think the next move for me will be to move some data loading to a non-reactive solution like Apollo. I've spent so much time fighting with Meteor and its opaque behavior in terms of resource usage and performance. It may save time to switch some data loading to a solution that has more manageable performance characteristics rather than trying to further optimize Meteor publications.

**Comment by KoenLav** [comment]
@AndrewMorsillo thanks for your insights!

I definitely don't think Meteor/DDP's design is flawed at its core, however it seems it could use some re-architecture.

**Comment by AndrewMorsillo** [comment]
I ended up remove all "reactive publishing" from my app and have gotten a huge boost in stability. Resource usage is still fairly high but much more consistent. I think the high resource usage is just the nature of Meteor publications at this point and not a bug.

**Comment by AndrewMorsillo** [comment]
I don't think it is flawed at its core either, just that it is more suitable for some jobs vs others and that isn't very clear. In my case publishing a lot of data in a complex app that is strictly per-user seems to bring out some of the less desirable performance characteristics of DDP. I think in a simpler app that also shared more data between clients that things would probably look at lot better.

**Comment by evolross** [comment]
Pub-sub definitely has overhead. For example, every single user's current published data set is kept in memory on the server despite cursor reuse. So if each user has megabytes of data in your case... that could add up.

**Comment by nick-gudumac** [comment]
@benjamn I've just tested the 1.8.1 beta 8 and it looks like there's no more CPU pegging like before. Though, I haven't seen any performance improvements from 1.8 version (with node 8.11.2). Also, the CPU still spikes if there are bulk updates in MongoDB (using Oplog), similar to previous versions.

**Comment by benjamn** [comment]
Thanks for that confirmation @nick-gudumac! I was just about to comment here asking folks to run
```bash
meteor update --release 1.8.1-beta.8
```
Note that we've moved two minor versions beyond Node 8.12.0 at this point, to Node 8.14.0!

**Comment by benjamn** [comment]
In order to reflect what's been fixed in the Meteor 1.8.1 milestone so far, I'm going to close this issue now, though of course you'll need to run
```bash
meteor update --release 1.8.1-beta.8
```
to get Node 8.14.0. If you're not using Galaxy, and you have the freedom to choose the Node version you're using in production independently of your Meteor version, you may be able to run just
```bash
meteor update promise
```
instead of updating to Meteor 1.8.1 (yet).

**Comment by pawansingh-snapwiz** [comment]
Is this already part of -- 1.8.1?
@benjamn Please let me know, I wan't to upgrade to 1.8.1, but getting confused with this - **so this version of Meteor should be considered experimental until those problems have been fixed**, which is written in 1.8.1 changelog page.

https://docs.meteor.com/changelog.html#changes-8

Note: While Node 8.12.0 included changes that may improve the performance of Meteor apps, there have been reports of CPU usage spikes in production due to excessive garbage collection, so this version of Meteor should be considered experimental until those problems have been fixed. Issue #10216

Thanks in advance folr your help.

**Comment by abernix** [comment]
@pawansingh-snapwiz While this was fixed in Meteor 1.8.1, the version of Node.js in Meteor 1.8.1 (Node.js 8) is no longer under long-term support by the Node.js Foundation and is not receiving security updates.  Please consider upgrading to a newer Meteor version which bundles a version still under [long-term support](https://github.com/nodejs/Release#release-schedule), e.g., Meteor 1.9+ (Though Meteor 1.10.2 is available and shipped with Node.js 12).
